File: sql/core/src/main/scala/org/apache/spark/sql/Arrow.scala
/*
* Licensed to the Apache Software Foundation (ASF) under one or more
* contributor license agreements.  See the NOTICE file distributed with
* this work for additional information regarding copyright ownership.
* The ASF licenses this file to You under the Apache License, Version 2.0
* (the "License"); you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

package org.apache.spark.sql

import scala.collection.JavaConverters._
import scala.language.implicitConversions

import io.netty.buffer.ArrowBuf
import org.apache.arrow.memory.{BaseAllocator, RootAllocator}
import org.apache.arrow.vector.{BaseDataValueVector, BitVector, IntVector, VarBinaryVector}
import org.apache.arrow.vector.BaseValueVector.BaseMutator
import org.apache.arrow.vector.schema.{ArrowFieldNode, ArrowRecordBatch}
import org.apache.arrow.vector.types.FloatingPointPrecision
import org.apache.arrow.vector.types.pojo.{ArrowType, Field, Schema}

import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.types._
import org.apache.spark.unsafe.types.UTF8String

object Arrow {

  private def sparkTypeToArrowType(dataType: DataType): ArrowType = {
    dataType match {
      case BooleanType => ArrowType.Bool.INSTANCE
      case ShortType => new ArrowType.Int(4 * ShortType.defaultSize, true) // TODO - check on this
      case IntegerType => new ArrowType.Int(8 * IntegerType.defaultSize, true)
      case LongType => new ArrowType.Int(8 * LongType.defaultSize, true)
      case FloatType => new ArrowType.FloatingPoint(FloatingPointPrecision.SINGLE)
      case DoubleType => new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)
      case ByteType => new ArrowType.Int(8, false)
      case StringType => ArrowType.Utf8.INSTANCE
      case _ => throw new UnsupportedOperationException(s"Unsupported data type: ${dataType}")
    }
  }

  private def schemaToColumnWriter(allocator: RootAllocator, dataType: DataType): ColumnWriter = {
    // TODO: Implement more types
    dataType match {
      case IntegerType => new IntegerColumnWriter(allocator)
      case StringType => new UTF8StringColumnWriter(allocator)
    }
  }

  /**
   * Transfer an array of InternalRow to an ArrowRecordBatch.
   */
  def internalRowsToArrowRecordBatch(
      rows: Array[InternalRow],
      schema: StructType,
      allocator: RootAllocator): ArrowRecordBatch = {
    val bufAndField = schema.fields.zipWithIndex.map { case (field, ordinal) =>
      internalRowToArrowBuf(rows, ordinal, field, allocator)
    }

    val fieldNodes = bufAndField.flatMap(_._1).toList.asJava
    val buffers = bufAndField.flatMap(_._2).toList.asJava

    new ArrowRecordBatch(rows.length, fieldNodes, buffers)
  }

  /**
   * Convert an array of InternalRow to an ArrowBuf.
   */
  def internalRowToArrowBuf(
      rows: Array[InternalRow],
      ordinal: Int,
      field: StructField,
      allocator: RootAllocator): (Array[ArrowFieldNode], Array[ArrowBuf]) = {
    val numOfRows = rows.length
    val columnWriter = schemaToColumnWriter(allocator, field.dataType)
    columnWriter.init(numOfRows)
    var index = 0

    while(index < numOfRows) {
      val data = rows(index).get(ordinal, field.dataType)
      columnWriter.write(data)
      index += 1
    }

    val (arrowFieldNodes, arrowBufs) = columnWriter.finish()
    (arrowFieldNodes.toArray, arrowBufs.toArray)

  }

  private[sql] def schemaToArrowSchema(schema: StructType): Schema = {
    val arrowFields = schema.fields.map(sparkFieldToArrowField)
    new Schema(arrowFields.toList.asJava)
  }

  private[sql] def sparkFieldToArrowField(sparkField: StructField): Field = {
    val name = sparkField.name
    val dataType = sparkField.dataType
    val nullable = sparkField.nullable
    new Field(name, nullable, sparkTypeToArrowType(dataType), List.empty[Field].asJava)
  }
}

trait ColumnWriter {
  def init(initialSize: Int): Unit
  def write(data: Any): Unit
  def finish(): (Seq[ArrowFieldNode], Seq[ArrowBuf])
}

/**
 * Base class for flat arrow column writer, i.e., column without children.
 */
abstract class PrimitiveColumnWriter(protected val allocator: BaseAllocator) extends ColumnWriter {
  protected val validityVector = new BitVector("validity", allocator)
  protected val validityMutator = validityVector.getMutator
  protected val valueVector: BaseDataValueVector
  protected val valueMutator: BaseMutator

  var count = 0
  var nullCount = 0

  protected def writeNull()
  protected def writeData(data: Any)
  protected def valueBufs(): Seq[ArrowBuf] = List(valueVector.getBuffer)

  override def init(initialSize: Int): Unit = {
    validityVector.allocateNew(initialSize)
    valueVector.allocateNew()
  }

  override def write(data: Any): Unit = {
    if (data == null) {
      validityMutator.setSafe(count, 0)
      writeNull()
      nullCount += 1
    } else {
      validityMutator.setSafe(count, 1)
      writeData(data)
    }

    count += 1
  }

  override def finish(): (Seq[ArrowFieldNode], Seq[ArrowBuf]) = {
    validityMutator.setValueCount(count)
    valueMutator.setValueCount(count)

    val fieldNode = new ArrowFieldNode(count, nullCount)
    (List(fieldNode), validityVector.getBuffer +: valueBufs)
  }
}

class IntegerColumnWriter(allocator: BaseAllocator) extends PrimitiveColumnWriter(allocator) {
  override protected val valueVector: IntVector = new IntVector("IntValue", allocator)
  override protected val valueMutator: IntVector#Mutator = valueVector.getMutator

  override protected def writeNull(): Unit = valueMutator.set(count, 0)
  override protected def writeData(data: Any): Unit
  = valueMutator.setSafe(count, data.asInstanceOf[Int])
}

class UTF8StringColumnWriter(allocator: BaseAllocator) extends PrimitiveColumnWriter(allocator) {
  override protected val valueVector: VarBinaryVector
  = new VarBinaryVector("UTF8StringValue", allocator)
  override protected val valueMutator: VarBinaryVector#Mutator = valueVector.getMutator

  override protected def writeNull(): Unit = valueMutator.setSafe(count, Array.empty[Byte])
  override protected def writeData(data: Any): Unit
  = valueMutator.setSafe(count, data.asInstanceOf[UTF8String].getBytes)
  override protected def valueBufs(): Seq[ArrowBuf]
  = List(valueVector.getOffsetVector.getBuffer, valueVector.getBuffer)
}

File: python/pyspark/arrow.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from pyspark.rdd import RDD
from pyspark.sql import DataFrame, GroupedData
from pyspark.serializers import ArrowPandasSerializer

'''
class DataFrame(object):
    ...
    def asPandas(self):
        return ArrowDataFrame(self)

def mapPartitionsAsPandas(self, f):
 +        """
 +        Return an Arrow RDD with function applied to a ``pandas.DataFrame`` at each partition.
 +        """
 +        def f_process_pandas(pdf_iter):
 +            for pdf in pdf_iter:
 +                yield f(pdf)
 +
 +        payload_jrdd = self._jdf.toArrowPayloadBytes().toJavaRDD()
 +        rdd = ArrowRDD(payload_jrdd, self._sc)
 +        return rdd.mapPartitions(f_process_pandas)
'''

class ArrowRDDBase(object):
    """
    Wraps a Python RDD to deserialize using Arrow into ``pandas.DataFrame`` for processing.
    """

    def __init__(self, sql_ctx):
        self._sql_ctx = sql_ctx

    @property
    def _rdd(self):
        raise NotImplementedError

    def _wrap_rdd(self, rdd):
        rdd._jrdd_deserializer = self._rdd._jrdd_deserializer
        return ArrowRDD(jrdd=None, sql_ctx=self._sql_ctx, pipelined_rdd=rdd)

    def map(self, f, preservesPartitioning=False):
        rdd = self._rdd.map(f, preservesPartitioning=preservesPartitioning)
        return self._wrap_rdd(rdd)

    def reduce(self, f):
        return self._rdd.reduce(f)

    def count(self):
        return self._rdd.count()

    def collect(self):
        return self._rdd.collect()


class ArrowDataFrame(ArrowRDDBase):
    """
    Wraps a Python DataFrame to group/winow then apply using``pandas.DataFrame``
    """

    def __init__(self, data_frame):
        super(ArrowDataFrame, self).__init__(data_frame.sql_ctx)
        self.df = data_frame
        self._lazy_rdd = None

    @property
    def _rdd(self):
        if self._lazy_rdd is None:
            payload_jrdd = self.df._jdf.toArrowPayloadBytes().toJavaRDD()
            self._lazy_rdd = ArrowRDD(payload_jrdd, self._sql_ctx)
        return self._lazy_rdd

    def groupBy(self, *cols):
        jgd = self._jdf.groupBy(self._jcols(*cols))
        return ArrowGroupedData(jgd, self.df.sql_ctx)

    def windowOver(self, window_spec):
        raise NotImplementedError()


class ArrowGroupedData(GroupedData):
    """
    Wraps a Python GroupedData object to process groups as ``pandas.DataFrame``
    """

    def __init__(self, jgd, sql_ctx):
        super(ArrowGroupedData, self).__init__(jgd, sql_ctx)

    def agg(self, f):
        # Apply function f to each group
        return DataFrame(...)


class ArrowRDD(ArrowRDDBase):
    """
    Wraps a Python RDD to deserialize using Arrow into ``pandas.DataFrame`` for processing.
    """

    def __init__(self, jrdd, sql_ctx, pipelined_rdd=None):
        super(ArrowRDD, self).__init__(sql_ctx)
        if pipelined_rdd is None:
            self._rdd_obj = RDD(jrdd, self._sql_ctx._sc, jrdd_deserializer=ArrowPandasSerializer())
        else:
            self._rdd_obj = pipelined_rdd

    @property
    def _rdd(self):
        return self._rdd_obj

    def toDF(self):
        schema = _parse_datatype_json_string(jschema.json())
        return self._sql_ctx.sparkSession.createDataFrame(self._rdd, schema=schema)
